src/ports/postgres/modules/knn/knn.py_in

```diff
@@ -124,16 +137,312 @@ def knn_validate_src(schema_madlib, point_source, point_column_name, point_id,
+    if depth <= 0:
+        plpy.error("kNN Error: depth={0} is an invalid value, must be greater "
+                   "than 0.".format(depth))
+    if leaf_nodes <= 0:
+        plpy.error("kNN Error: leaf_nodes={0} is an invalid value, must be greater "
+                   "than 0.".format(leaf_nodes))
+    if pow(2,depth) <= leaf_nodes:
+        plpy.error("kNN Error: depth={0}, leaf_nodes={1} is not valid. "
+                   "The leaf_nodes value must be lower than 2^depth".format(depth, leaf_nodes))
     return k
 # ------------------------------------------------------------------------------
 
 
+def kd_tree(schema_madlib, source_table, output_table, point_column_name, depth,
+            r_id, dim, **kwargs):
+    """
+        KD-tree function to create a partitioning for KNN
+        Args:
+            @param schema_madlib        Name of the Madlib Schema
+            @param source_table         Training data table
+            @param output_table         Name of the table to store kd tree
+            @param point_column_name    Name of the column with training data
+                                        or expression that evaluates to a
+                                        numeric array
+            @param depth                Depth of the kd tree
+            @param r_id                 Name of the region id column
+            @param dim                  Name of the dimension column
+
+    """
+    with MinWarning("error"):
+
+        validate_kd_tree(source_table, output_table, point_column_name, depth)
+        n_features = num_features(source_table, point_column_name)
+
+        clauses = [" WHERE 1=1 "]
+        cutoffs = []
+        centers_table = add_postfix(output_table, "_centers")
+        clause_counter = 0
+        current_feature = 1
+        for curr_level in range(depth):
+            for curr_leaf in range(pow(2,curr_level)):
+                clause = clauses[clause_counter]
+                cutoff_sql = """
+                    SELECT percentile_disc(0.5)
+                           WITHIN GROUP (
+                            ORDER BY ({point_column_name})[{current_feature}]
+                           ) AS cutoff
+                    FROM {source_table}
+                    {clause}
+                    """.format(**locals())
+
+                cutoff = plpy.execute(cutoff_sql)[0]['cutoff']
+                cutoff = cutoff if cutoff is not None else "NULL"
+                clause_counter += 1
+
+                cutoffs.append(cutoff)
+                clauses.append(clause +
+                               "AND ({point_column_name})[{current_feature}]"
+                               " < {cutoff} ".format(**locals()))
+                clauses.append(clause +
+                               "AND ({point_column_name})[{current_feature}]"
+                               " >= {cutoff} ".format(**locals()))
+            current_feature = current_feature % n_features + 1
```

Comment:
I prefer `curr_feature = curr_level % n_features + 1` placed under the `curr_level` for loop.